rewrite-java/src/main/java/org/openrewrite/java/recipes/MigrateRecipeToRewrite8.java

```diff
@@ -554,20 +553,46 @@ private static J.MethodInvocation getPreconditionsNotTemplate() {
     @SuppressWarnings("all")
     private static J.MemberReference getVisitMemberReferenceTemplate() {
         if (visitMemberReferenceTemplate == null) {
-            visitMemberReferenceTemplate = PartProvider.buildPart(
-                "import org.openrewrite.java.JavaVisitor;\n" +
-                "import org.openrewrite.java.tree.J;\n" +
-                "import org.openrewrite.java.tree.JavaSourceFile;\n" +
-                "public class A<P> extends JavaVisitor<P> {\n" +
-                "    @Override\n" +
-                "    public J visitJavaSourceFile(JavaSourceFile cu, P p) {\n" +
-                "        return visitAndCast(cu, p, super::visit);\n" +
-                "    }\n" +
-                "}", J.MemberReference.class,
-                JavaParser.runtimeClasspath()
-                );
+            visitMemberReferenceTemplate = parseAndBuild(
+                    "import org.openrewrite.java.JavaVisitor;\n" +
+                    "import org.openrewrite.java.tree.J;\n" +
+                    "import org.openrewrite.java.tree.JavaSourceFile;\n" +
+                    "public class A<P> extends JavaVisitor<P> {\n" +
+                    "    @Override\n" +
+                    "    public J visitJavaSourceFile(JavaSourceFile cu, P p) {\n" +
+                    "        return visitAndCast(cu, p, super::visit);\n" +
+                    "    }\n" +
+                    "}", J.MemberReference.class,
+                    JavaParser.runtimeClasspath()
+            );
         }
 
         return visitMemberReferenceTemplate;
     }
+
+    private static <J2 extends J> J2 parseAndBuild(@Language("java") String code,
+                                              Class<J2> expected,
+                                              Collection<Path> classpath) {
+        JavaParser.Builder<? extends JavaParser, ?> builder = JavaParser.fromJavaVersion().classpath(classpath);
+
+        J.CompilationUnit cu = builder.build()
+            .parse(code)
+            .findFirst()
+            .orElseThrow(() -> new IllegalArgumentException("Could not parse as Java"));
+        List<J2> parts = new ArrayList<>(1);
+        new JavaVisitor<List<J2>>() {
+            @Override
+            public @Nullable J visit(@Nullable Tree tree, List<J2> j2s) {
+                if (expected.isInstance(tree)) {
+                    //noinspection unchecked
+                    J2 j2 = (J2) tree;
+                    j2s.add(j2);
+                    return j2;
+                }
+                return super.visit(tree, j2s);
+            }
+        }.visit(cu, parts);
+
+        return parts.get(0);
+    }
```

Comment:
Any reason why not to make this a public API? This could be incredibly useful for synthesizing types.

`rewrite-security` already has a utility to do something very similar:

https://github.com/openrewrite/rewrite-java-security/blob/73dbecaf9354756d26a135766dd044559f6cf6f1/src/main/java/org/openrewrite/java/security/internal/TypeGenerator.java#L31-L52